typing/typedecl.ml

```diff
@@ -666,32 +679,29 @@ let check_well_founded env loc path to_check ty =
       | Tobject _ | Tvariant _ -> true
       | _ -> !Clflags.recursive_types
     in
-    let visited' = TypeMap.add ty parents !visited in
-    let arg_exn =
-      try
-        visited := visited';
-        let parents =
-          if rec_ok then TypeSet.empty else TypeSet.add ty parents in
-        Btype.iter_type_expr (check ty0 parents) ty;
-        None
-      with e ->
-        visited := visited'; Some e
-    in
+    let ty0 = if TypeSet.is_empty parents then ty else ty0 in
+    if rec_ok then Btype.iter_type_expr (check false ty0 TypeSet.empty) ty else
+    let parents = TypeSet.add ty parents in
     match get_desc ty with
-    | Tconstr(p, _, _) when arg_exn <> None || to_check p ->
-        if to_check p then Option.iter raise arg_exn
-        else Btype.iter_type_expr (check ty0 TypeSet.empty) ty;
-        begin try
-          let ty' = Ctype.try_expand_once_opt env ty in
-          let ty0 = if TypeSet.is_empty parents then ty else ty0 in
-          check ty0 (TypeSet.add ty parents) ty'
-        with
-          Ctype.Cannot_expand -> Option.iter raise arg_exn
+    | Tconstr(p, tyl, _) ->
+        let to_check = to_check p in
+        if to_check then List.iter (check root ty0 parents) tyl
+        else List.iter (check false ty0 TypeSet.empty) tyl;
+        (* Only expand nodes on a root path, or if we need to check
+           whether some parameters are accessible *)
+        if root || not to_check && tyl <> [] then begin
+          try
+            let ty' = Ctype.try_expand_once_opt env ty in
+            check root ty0 (TypeSet.add ty parents) ty'
+          with
+            Ctype.Cannot_expand ->
+              List.iter (check root ty0 parents) tyl
```

Comment:
Could this be written as the more precise

```ocaml
match Ctype.try_expand_once_opt env ty with
| ty' -> ...
| exception Ctype.Cannot_expand -> ...
```

Generally in the type-checker code the use of `try..with` makes it hard to trace which subterm we expect to raise an exception.

Reply:
Done